### PR TITLE
Add JS validation for MCQ/MSQ for number of options in instructorFeedbackEditPage #2765

### LINKED ISSUE #2765: Add JS validation for MCQ/MSQ for number of options in instructorFeedbackEditPage
Currently, there is a server side check to ensure at least 2 options for MCQ/MSQ.
We can add in client side JS validation to prevent the user from having less than 2 options (by preventing deletion of options if there are two or less options).
This will also ensure the UI remains consistent and not have times where there are no options showing.

### PR DESCRIPTION
**Pull request by MrFool**
Fixes #2765

### CODE REVIEW

**Comment by MrFool** [comment]
i'll update when it's green, no need to set to review yet

File: src/main/webapp/js/instructorFeedbackEdit.js [hunk 1]
```diff
@@ -805,17 +805,27 @@ function addMcqOption(questionNumber) {
...
+	    idSuffix = '--1';
+	}
+
+	var $thisRow = $('#mcqOptionRow-' + index + idSuffix);
+
+	// count number of child rows the table have and - 1 because of add option button
+	var numberOfOptions = $thisRow.parent().children('div').length - 1;
+
+	if (numberOfOptions <= 2) {
+	  alert('Too little choices for Multiple-choice (single answer) question. Minimum number of options is: 2.');
```
Comment (LowWeiLin) [inline src/main/webapp/js/instructorFeedbackEdit.js]: Using alerts is probably a bad idea.
When the user tries to remove an option and there are <= 2 options, he is probably trying to clear the text and write something else. Maybe do that instead?

**Comment by MrFool** [comment]
Sounds good, didn't think of it that way, Let me fix that along with the tests in awhile.

In that case, no need to tell the user why the action is not allowed - now that it is no longer possible to have < 2 options, there is no longer a way to see the backend generated error message.

**Comment by MrFool** [comment]
Removed alert, test cases green, with some edits to rubric test cases because it tried to delete past 2 rows to 1 row at some points.

Still thinking about how to show the "can't be done" message if we clear the text box after clicking and there is less than 2 rows, thoughts?

**Comment by MrFool** [comment]
By the way, 

screen shot 2015-04-01 at 3 51 58 pm

screen shot 2015-04-01 at 3 53 38 pm


MCQ using MSQ HTML, intended or? If it's not intended, should open new issue?

I fixed the JS to work because one of them was being attached the wrong deletion event handler, but now it's confusing because their HTML has been intertwined before I started on the issue.

**Comment by LowWeiLin** [comment]
Isn't that correct? the top one is Multiple-choice (multiple answers) which is MSQ, while the bottom one is Multiple choice (single answer) which is MCQ.

**Comment by LowWeiLin** [comment]
In that case, no need to tell the user why the action is not allowed - now that it is no longer possible to have < 2 options, there is no longer a way to see the backend generated error message.
>> Backend messages can still be seen if the user meddles with the javascript. Also, it might be better to only clear the fields if there is only one option left, as users might want to re-order the options.

Still thinking about how to show the "can't be done" message if we clear the text box after clicking and there is less than 2 rows, thoughts?
>> Probably no need to show any message. It should be enough that something happens(text gets cleared)

**Comment by MrFool** [comment]
Right, I messed up MCQ and MSQ's naming.

I got one more question though - if a user can't have less than 2 options anymore, how would we get to 'only one option'?

**Comment by LowWeiLin** [comment]
I'm trying to say that the minimum number of options on the UI should be 1 (but back-end still checks for min of 2). The main point of this issue is to prevent 0 options on the UI, which looks confusing.

**Comment by MrFool** [comment]
I see I see, understood, how about for rubric? When I forked it, it says check for 2 and 2, is it still 2 and 2, or 1 by 1 or?

**Comment by MrFool** [comment]
by "it", I mean the comments that were marked:

//TODO: Check for 2

etc.

**Comment by LowWeiLin** [comment]
Same as mcq, msq, there should be at least one cell(1 row, one col)

**Comment by MrFool** [comment]
Done, now checks for <= 1 for MSQ MCQ and Rubric

JS added html for MSQ options also now correctly calls removeMsqOption rather than removeMcqOption previously

**Comment by MrFool** [comment]
Just ran full test suite, dev green

**Comment by LowWeiLin** [comment]
Nice. Could you also add/modify the UI tests to check for this new behaviour?

**Comment by MrFool** [comment]
Tests added

**Comment by MrFool** [comment]
Ok so for a quick explanation of the changes -

1. removeMcqOption / removeMsqOption - variables were declared without var keyword, added the var keyword.

2. removeMcqOption / removeMsqOption - was using == instead of === which is not recommended in JavaScript due to type coercion, fixed to use ===.

3. removeMcqOption / removeMsqOption - added checks for whether one last option is left and if so, they now clear the text field instead of removing the option entirely which leads to an MCQ / MSQ question not having an options

4. JavaScript generated MSQ HTML was incorrectly calling removeMcqOption, now fixed to call the correct removeMsqOption.

5. Made the same changes of 1-3 above for removing rubric row and rubric column.

6. Added tests for checking that the above functionalities added are not violated and passing in FeedbackRubricQuestionUiTest.java and InstructorFeedbackEditPageUiTest.java.

7. In InstructorFeedbackEditPage.java - 
  clickRemoveRubricRowLinkAndConfirm had a WebElement named addRubricRowLink -

  fixed to be named removeRubricRowLink to avoid confusion

  and

  clickRemoveRubricColLinkAndConfirm had a WebElement named addRubricColLink -

  fixed to be named removeRubricColLink to avoid confusion

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java [hunk 2]
```diff
@@ -329,7 +329,22 @@ private void testInputValidationForMcqQuestion() {
...
+        feedbackEditPage.clickNewQuestionButton();
+        feedbackEditPage.fillQuestionBox("Test question text");
+        assertTrue(feedbackEditPage.verifyNewMcqQuestionFormIsDisplayed());
+
+        feedbackEditPage.clickRemoveMcqOptionLink(1, -1);
+        assertEquals(false, feedbackEditPage.isElementPresent("mcqOptionRow-1--1"));
         
+        assertEquals(true, feedbackEditPage.isElementPresent("mcqOptionRow-0--1"));
+        feedbackEditPage.clickRemoveMcqOptionLink(0, -1);
+        assertEquals(true, feedbackEditPage.isElementPresent("mcqOptionRow-0--1"));
```
Comment (LowWeiLin) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: Also check that the value is empty?

File: src/test/java/teammates/test/cases/ui/browsertests/FeedbackRubricQuestionUiTest.java [hunk 3]
```diff
@@ -304,6 +305,33 @@ private void testDeleteRubricQuestionAction() {
         assertNull(BackDoor.getFeedbackQuestion(courseId, feedbackSessionName, 1));    
     }
     
+    private void testInputJsValidationForRubricQuestion() {
+        ______TS("JS validation test");
```
Comment (LowWeiLin) [inline src/test/java/teammates/test/cases/ui/browsertests/FeedbackRubricQuestionUiTest.java]: Add comments to make it easier to understand what is being done in the test, and what is tested.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java (hunk 2 above)
Comment (LowWeiLin) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorFeedbackEditPageUiTest.java]: For the other question types too.

**Comment by LowWeiLin** [comment]
Thanks for the summary! Added some comments on the testing, after that it should be fine.

**Comment by MrFool** [comment]
Added comments to the testing, unable to test for strings unless some overhaul is made - imo should be a new issue:

1. For the MCQs the value of the input box does not update when we type things into the field - remains as empty string even after selenium adds "choice 1" to it.

2. For Rubric, the textarea's inner HTML remains as the default until we type in new things and we click "submit"

**Comment by LowWeiLin** [comment]
Why not able to test the strings? Have you tried creating a function similar to getMaxNumScaleBox in InstructorFeedbackEditPage.java? Seems weird that selenium does not support this.
If it can't be done, place a TODO comment there instead.

**Comment by MrFool** [comment]
Yupp, I tried to target the webElement and do a .getText(), so the textarea's text was still the default despite it being empty on the frontend side - but after I manually click save question, it was indeed the new text that was typed in from the UI. It's not selenium's problem, it's got to do with the fact that the textarea's innerHTML does not update when the user changes the value inside the textarea.

**Comment by LowWeiLin** [comment]
Try .getAttribute("innerText") or .getAttributes("textContent") instead of .getText()? From a quick search, some people seem to use that as an alternative. Otherwise, its fine to just add a TODO comment.

**Comment by MrFool** [comment]
Yupp, just tested those as well, they also get the same value as getText() and since the inner text does not change when text is changed by the users, it returns the original value of

"This student participates well in online discussions."

E.g.

screen shot 2015-04-05 at 1 06 17 am
screen shot 2015-04-05 at 1 06 01 am


and for MCQ/MSQ, the original inner text / html is empty, it remains empty.

One way we can do the checks would be to bind the text inside the box with the HTML, but that requires more JavaScript than we have at the moment. We could look into that by opening another issue? But I'd think that is a major change.

**Comment by MrFool** [comment]
I can open an issue and take a deeper look into it as to why the values / inner text is not being updated when the user changes the values in the text box / text area

**Comment by LowWeiLin** [comment]
Okay sure, go ahead and create the issue.

**Comment by LowWeiLin** [comment]
Just realized that Distribute points (among recipients) question type also needs this. Will create a separate issue for it.

**Comment by MrFool** [comment]
Hmm ok 3262 shouldn't be too different from this, I'll see if I can tackle that, but 3263 might be harder considering that binding the text input to the HTML might be a lot more of an undertaking

**Pull request merged**